pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -3402,6 +3488,454 @@ def SeismicActivityExceptionRule(model, k, t):
+            mh.s_L - mh.s_P,
+            mh.s_T,
+            rule=MAOPressureRule,
+            doc="Max allowable pressure rule",
+        )
+
+        mh.objective = Objective(
+            expr=(
+                mh.v_Z
+                + (
+                    sum(
+                        sum(
+                            (mh.v_PumpHead[key, t] * 10 + mh.v_ValveHead[key, t])
+                            for key in mh.s_LLA
+                        )
+                        for t in mh.s_T
+                    )
+                )
+            ),
+            sense=minimize,
+            doc="Objective function",
+        )
+
+        # MINLP solver is currently accessed through GAMS. User must specify
+        # available solvers to solve this problem.
+        # TODO: check to use open source solvers
+        solver = SolverFactory("gams")
+        mathoptsolver = "baron"
+        solver_options = {
+            "firstfeas": 1,
+        }
+
+        if not os.path.exists("temp"):
+            os.makedirs("temp")
+
+        with open("temp/" + mathoptsolver + ".opt", "w") as f:
+            for k, v in solver_options.items():
+                f.write(str(k) + " " + str(v) + "\n")
+
+        solver.solve(
+            mh,
+            tee=True,
+            keepfiles=True,
+            solver=mathoptsolver,
+            tmpdir="temp",
+            add_options=["gams_model.optfile=1;"],
+        )
+
+        # second solve with dicopt
+        solver2 = SolverFactory("gams")
+        mathoptsolver2 = "dicopt"
+        solver_options2 = {
+            "tol": 1e-3,
+            "max_iter": 1000,
+            "constr_viol_tol": 0.009,
+            "acceptable_constr_viol_tol": 0.01,
+            "acceptable_tol": 1e-6,
+            "mu_strategy": "adaptive",
+            "mu_init": 1e-10,
+            "mu_max": 1e-1,
+            "print_user_options": "yes",
+            "warm_start_init_point": "yes",
+            "warm_start_mult_bound_push": 1e-60,
+            "warm_start_bound_push": 1e-60,
+        }
+
+        if not os.path.exists("temp"):
+            os.makedirs("temp")
+
+        with open("temp/" + mathoptsolver2 + ".opt", "w") as f:
+            for k, v in solver_options2.items():
+                f.write(str(k) + " " + str(v) + "\n")
+
+        solver2.solve(
+            mh,
+            tee=True,
+            keepfiles=True,
+            solver=mathoptsolver2,
+            tmpdir="temp",
+            add_options=["gams_model.optfile=1;"],
+        )
+
+    return mh
+
+
+def _hazen_williams_head(mat_factor, length, diameter, flow):
+    """
+    Computes Hazen Williams (HW) head in a pipeline
+
+    Input Args
+    ----------
+    mat_factor: pipeline material factor for HW equation
+    length : length of pipeline segment in m.
+    diameter : diameter of pipeline segment in m.
+    flow : water flow through the pipeline segment in m3/s.
+
+    Returns
+    -------
+    hw_friction_head : TYPE
+        DESCRIPTION.
```

Comment:
This documentation for the return value should be cleaned up.

Reply:
done